[hailtop] add RouterAsyncFS.copy

cseed opened the pull request:
This PR is the first iteration of an AsyncFS-based copy interface.  It adds RouterAsyncFS.copy.

The goal of these changes is to establish the interface and behavior.  I expect several follow-on PRs:

 - Revise the original copy interface proposal and add to dev-docs.
 - Parallelizes the transfers concurrently with async and across multiple threads.
 - After parallizing, copy will involve a lot of paralellism.  Throwing an exception on the first error will be very non-deterministic.  Instead, copy will return a report that collects all the errors that were encountered in the course of copying, and summarizes how many files/bytes were copied.
 - Avoid overwriting the destination if it exists and has a matching checksum (or size).
 - Introduce multi-part transfers add a post-pass for Google Storage to detect file-and-directory errors.
 - Adds support for S3.
 -  Add `hailctl cp ...`.
 - Use copy in Batch.

After this goes in, these can mostly be developed in parallel.

A few principles guided the implementation of copy: perform the minimal number of system calls or API requests per copy, and only do error checking when it doesn't involve additional FS operations.  For example, it is too expensive to exhaustively check if we're creating a path that is a file and a directory in Google Storage.

I considered doing additional and exhaustive checking for the actual copy arguments.  For example, currently, `cp -T /path/to/file /path/to/dir` will not generate an error on Google Storage.  In the end, I decided to go with the current behavior and I will add an option to do a postpass to check for file-and-dir paths.

To achieve this, for each transfer, I simultaneously stat the destination (if needd) to determine if it is a file, directory or doesn't exist.  For each source, I simultaneously try to copy it as a file and a directory.  When copying each source, we don't need to know the type of the destination until after we've stat'ed the source, so stat'ing the sources and destinations are all overlapping.  This avoids dependencies where I have to e.g. stat the input, decide what to do, and then perform a second action.

I approached testing two ways: First, hand test common operations and errors (copy file, copy dir, overwrite, overwrite dir with file and vice versa, the various treat_dest_as settings, large files, detecting copy-and-files on input on Google Storage, etc.)  Second, I enumerated essentially all single input transfers and recorded the results.  These are then tested on every pair of file systems: file to file, file to gs, gs to file, etc.  To make this run reasonably, I parallelize using xdist and the full tests (1364 tests) now take about 2 minutes.  (If you restrict to the just the local filesystem, the tests just take a couple seconds.)  The second test verifies that all cases run without unexpected errors, and that the behavior is file system independent.

jigold commented on hail/python/test/hailtop/aiotools/generate_copy_test_specs.py:
```diff
@@ -0,0 +1,139 @@
+    await fs.mkdir(f'{base}{path}')
+    if name == 'src':
+        await create_test_file(fs, name, base, f'{path}file1')
+    await fs.mkdir(f'{base}{path}subdir/')
```
Why does it matter whether the name is src or dest? Shouldn't the test dir always be the same?

cseed replied:
I made a clarifying comment.

jigold reviewed (changes requested):
I thoroughly looked at the actual module code. I started reading through the test code. There's a lot there. Is there something in particular you want me to be on the lookout for?

cseed commented on build.yaml:
```diff
@@ -1199,6 +1199,36 @@ steps:
+     set -ex
+     cd /io
+     tar xzf test.tar.gz
+     tar xvf debug-wheel-container.tar
```
Because we're testing code in hailtop.aiotools, so it needs to be installed so the test can call it.

cseed commented on hail/python/hailtop/aiotools/fs.py:
```diff
@@ -130,7 +146,10 @@ def name(self) -> str:
     async def url(self) -> str:
         trailing_slash = "/" if await self.is_dir() else ""
-        return f'{self._base_url}{self._entry.name}{trailing_slash}'
+        return f'{self.url_maybe_trailing_slash()}{trailing_slash}'
```
This was a mistake (but I think it worked for the reason you mentioned).  Fixed.

When this is all done I'm going to make a post pass and see if I can simplify the trailing / handling.

cseed commented on hail/python/hailtop/aiotools/fs.py:
```diff
@@ -230,6 +260,229 @@ async def rmtree(self, url: str) -> None:
+            if dest_type == AsyncFS.FILE:
+                raise NotADirectoryError(self.dest)
+
+            return url_join(self.dest, url_basename(self.src.rstrip('/'))), None
```
Yep!  In an earlier version, didn't unconditionally test the destination type (when it had a trailing slash).

cseed replied:
Ah, sorry, the original code is right.  We know dest is a directory at this point, but we're not copying to dest, we're copying to dest/basename(src), and we don't know it's type.  I added a comment to this effect.

cseed commented on hail/python/hailtop/aiotools/fs.py:
```diff
@@ -230,6 +260,229 @@ async def rmtree(self, url: str) -> None:
+            srcentries = await self.router_fs.listfiles(src, recursive=True)
+        except (NotADirectoryError, FileNotFoundError):
+            self.src_is_dir = False
+            await self.release_barrier()
```
Renamed `release_barrier_and_wait`.

cseed commented on hail/python/hailtop/aiotools/fs.py:
```diff
@@ -230,6 +260,229 @@ async def rmtree(self, url: str) -> None:
+            return_exceptions=True)
+
+        assert self.pending == 0
+        assert (self.src_is_file is None) == self.src.endswith('/')
```
Nope.  If the src has a trailing slash, we don't check its type, so src_is_file stays None.

cseed commented on hail/python/hailtop/aiotools/fs.py:
```diff
@@ -230,6 +260,229 @@ async def rmtree(self, url: str) -> None:
+            raise FileNotFoundError(self.src)
+        for result in results:
+            if isinstance(result, Exception):
+                raise result
```
Totally agree!  From my list of future work:

 > After parallizing, copy will involve a lot of paralellism. Throwing an exception on the first error will be very non-deterministic. Instead, copy will return a report that collects all the errors that were encountered in the course of copying, and summarizes how many files/bytes were copied.

I will print this report in the command line tool (`hailctl cp ...`)

jigold commented:
I'm still looking at this. I'm having a hard time with the test code. I'll try and spend more time tomorrow so I can either figure out what's going on myself or have specific questions.

jigold commented on hail/python/test/hailtop/aiotools/test_copy.py:
```diff
@@ -0,0 +1,351 @@
+
+    await fs.copy(Transfer(f'{src_base}a', dest_base.rstrip('/')))
+
+    await expect_file(fs, f'{dest_base}a/file1', 'src/a/file1')
```
Is there ever a case where we're expecting to get 

```
{dest_base}file1
{dest_base}subdir
```

Basically putting the contents of a/ directly inside {dest_base}?

cseed replied:
I added a test_copy_src_parts test for this that copies from multiple sources.  I think this would normally be handled by globbing which would run a layer above the copy function.

jigold commented on hail/python/test/hailtop/aiotools/test_copy.py:
```diff
@@ -0,0 +1,351 @@
+
+    await create_test_file(fs, 'src', src_base, 'a')
+
+    await fs.copy(Transfer(f'{src_base}a', f'{dest_base}x'))
```
What happens if this is `Target.TARGET_DIRECTORY` and `x` doesn't exist?

cseed replied:
OK, this is an interesting example and it is possible we want to change the behavior.  I added a test test_copy_file_dest_target_directory_doesnt_exist to test this.  It works, in which I mean it creates dest/x/a on both the local filesystem and gs.  This is because (1) if you set treat_dest_as=Transfer.TARGET_DIR, it doesn't actually stat the destination and just treats it as a directory.  Then when we go to copy src/a to dest/x/a, SourceCopier._copy_file creates the destination if it doesn't exist.  This is because when you copy cp dir dest and dir/a/b/c exists, we need to create dest/a/b.  In this case, dest itself doesn't exist, so we create it.  I'm not unhappy with the behavior, but it may be surprising since cp throws an error.

Either way, it won't throw an error on gs.

Options going forward:
 - Leave behavior as-is.
 - Always stat dest and give an error if it doesn't exist.  This will do unnecessary work for errors, which I've avoided so far.
 - Don't create the dest itself in _copy_file.  Then we'd get a FileNotFoundError on file, and no error on gs.

jigold commented on hail/python/test/hailtop/aiotools/test_copy.py:
```diff
@@ -0,0 +1,351 @@
+
+    await create_test_file(fs, 'src', src_base, 'a')
+
+    await fs.copy(Transfer(f'{src_base}a', dest_base, treat_dest_as=Transfer.TARGET_DIR))
```
Will this error if it's `Transfer.TARGET_FILE` instead?

cseed replied:
I added a test_copy_dest_target_file_is_dir for this case.  It throws an IsADirectoryError currently.

We might want to change this behavior to be more consistent and aggressive.  I notice that in Copier._dest_type, I don't check the dest if it has a trailing slash or treat_dest_as == Transfer.TARGET_DIR.  But I do stat the dest if treat_dest_as == Transfer.TARGET_FILE which seems unnecessary.

If I remove this check, which has slightly higher performance, we would write the destination getting either a IsADirectoryError or creating an undetected file-and-directory situation on Google Storage.

jigold reviewed (changes requested):
I tried to be devil's advocate and find any holes in the tests. You might have already handled those cases.

danking commented:
Cotton, bump, really looking forward to speedier file transfer!

cseed commented on hail/python/test/hailtop/aiotools/test_copy.py:
```diff
@@ -0,0 +1,351 @@
+    await create_test_file(fs, 'src', src_base, 'a')
+    await create_test_file(fs, 'src', src_base, 'b')
+
+    await fs.copy(Transfer([f'{src_base}a', f'{src_base}b'], dest_base.rstrip('/')))
```
> you need a lock on the file

What do you mean by lock?

jigold replied:
I mean that you don't want to be writing to the same file from two different sources in parallel. This is in the future. I just remembered that case when I was looking at the test.

cseed commented:
OK, I think I addressed all the comments.  I'm going to make the change to not stat the destination if we set treat_dest_as=Transfer.TARGET_FILE.

Then we just need to decide if we want to throw an error on file:// for a non-existent Transfer.TARGET_DIR.  I say no.

cseed commented:
OK, I changed _dest_type not to stat the destination if treat_dest_as=Transfer.FILE.  I regenerated the copy specs (a few changed) and the test_copy_dest_target_file_is_dir test only fails on the local filesystem.  Back to you.

jigold commented on hail/python/test/hailtop/aiotools/test_copy.py:
```diff
@@ -267,7 +267,7 @@ async def test_copy_dest_target_file_is_dir(copy_test_context):
     await create_test_file(fs, 'src', src_base, 'a')
 
-    with pytest.raises(IsADirectoryError):
+    with RaisesOrGS(dest_base, IsADirectoryError):
```
What's this change for?

jigold commented:
> Then we just need to decide if we want to throw an error on file:// for a non-existent Transfer.TARGET_DIR. I say no.

I agree. I think the user is being explicit what they intend to have happen.

jigold commented:
You have a pylint error with trailing whitespace.
